docs/proposals/24-shoot-oidc-issuer.md

```diff
@@ -0,0 +1,130 @@
+### Non-Goals
+ - Establish trust relations between clusters. This GEP is about simplifying the prerequisites for doing that.
+
+## Proposal
+
+This proposal consists of three parts that together will achieve the mentioned goals.
+
+### Static validation
+
+Static validation rule will be introduced enforcing that `shoot.Spec.Kubernetes.KubeAPIServer.ServiceAccountConfig.Issuer` is not set when the shoot is annotated with `gardener.cloud/managed-issuer: true`. This annotation will be incompatible with workerless shoots, as they do not have nodes and they do not run any workloads. The rule will also assure that once present the annotation cannot be removed from a shoot.
+
+```go
+if oldShoot.Annotations["gardener.cloud/managed-issuer"] == "true" {
+    // ensure that the new shoot also has this annotation
+    // ensure that newShoot.Spec.Kubernetes.KubeAPIServer.ServiceAccountConfig.Issuer is not set
+    // ensure that the shoot is not configured as workerless
+    newShoot.Annotations["gardener.cloud/managed-issuer"] = "true"
+} else if newShoot.Annotations["gardener.cloud/managed-issuer"] == "true" {
+    // ensure that newShoot.Spec.Kubernetes.KubeAPIServer.ServiceAccountConfig.Issuer is not set
+    // ensure that the shoot is not configured as workerless
+}
+```
+
+### Gardenlet
+
+Once annotated shoots will be easily identifiable by the `gardenlet`. If a shoot requires a managed issuer then the `gardenlet` will configure the `--service-account-issuer` and `--service-account-jwks-uri` flags for the shoot's [kube-apiserver](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/) deployment during reconciliation and set their values to `https://<central.domain.name>/projects/<project-name>/shoots/<shoot-uid>/issuer` and `https://<central.domain.name>/projects/<project-name>/shoots/<shoot-uid>/issuer/jwks` respectively (the central domain name should be configurable via the gardenlet's configuration). At the end of a shoot reconciliation flow, the `gardenlet` will fetch the OIDC metadata discovery documents from a shoot's `kube-apiserver` and sync them back to the Garden cluster in the form of a configmap following the naming convention `projectname--shootuid` in a special purposed namespace called `gardener-shoot-issuer`(name should be configurable). The [Seed Authorizer](https://github.com/gardener/gardener/blob/master/docs/deployment/gardenlet_api_access.md#scoped-api-access-for-gardenlets-and-extensions) will be enhanced so that Gardenlet is only permitted to perform actions on related configmaps. An additional advertised address will be added to the shoot status so that the issuer of the cluster is easily discoverable.
+
+```yaml
+status:
+  advertisedAddresses:
+    ...
+    - name: service-account-issuer
+      url: https://<central.domain.name>/projects/<project-name>/shoots/<shoot-uid>/issuer
+```
+
+During the deletion flow of a shoot the previously synced configmap will be deleted from the Garden cluster.
```

Comment:
Is explicit code needed here? I guess we could leverage owner references?

Reply:
I think it is. Cross-namespace owner references are disallowed by design https://kubernetes.io/docs/concepts/architecture/garbage-collection/#owners-dependents

Reply:
Oh, I missed that. Why don't we deploy this `ConfigMap into the namespace of the shoot? This is also where we put other shoot-related information like CA bundle or observability endpoints.

Reply:
There are several reasons that I do not suggest copying the configmap directly in the shoot namespace.

1. The shoot metadata server will need to have access to configmaps in all namespaces instead of one
2. The configmaps could be modified by users. Even if they are immutable, they can be deleted and recreated meaning that users will be able to control what the metadata server serves. This is particularly undesirable from my point of view since a user with malicious intents and access to the configmaps can for example self generate a signing key, then put the corresponding public key in the configmap and sign tokens that will be forwarded to external systems. And these external systems will think that these are actually tokens signed by the `kube-apiserver` and will give them access, when in fact these tokens are actually signed by a malicious actor.

Reply:
Thanks, I think the second point makes sense. Can you please add this information/motivation to the document so that it's clear why a dedicated namespace is getting used? Btw, who creates this namespace?

Reply:
Yes, I will add this information to the proposal as additional motivation.

> Btw, who creates this namespace?

I was thinking that the gardenlet can create this namespace. I am not sure when this will happen, but I imagine that it can be during startup of the gardenlet or during the shoot reconcilation, i.e. it will check that the namespace is created and if not - create it. Do you have any suggestions/ideas w.r.t this topic?

Reply:
There are multiple active gardenlets, one per seed, that would be responsible for the namespace and this could cause some sporadic issues. What is creating the `garden` namespace today, I think it should be able to create also this namespace, wdyt?

Reply:
Well, given that this namespace seems to be similar to the `gardener-system-seed-lease` namespace, I would create it here: https://github.com/gardener/gardener/blob/c29cc0fe30c09b980cb81d4a1caea6b7bccad244/cmd/gardener-apiserver/app/gardener_apiserver.go#L330-L335

Btw, let's appropriately call this namespace `gardener-system-shoot-issuers` then.

Reply:
I am 100% OK with the idea. I did not know that such namespace exist. I think that it would be even easier to have a constant name and not support name configuration. I will update the proposal with this suggestion.